test/test_futures.py

```diff
@@ -4,13 +4,49 @@
 import unittest
 from torch.futures import Future
 from torch.testing._internal.common_utils import IS_WINDOWS, TestCase, TemporaryFileName, run_tests
+from typing import TypeVar
+
+T = TypeVar("T")
 
 
 def add_one(fut):
     return fut.wait() + 1
 
 
 class TestFuture(TestCase):
+    def test_set_exception(self) -> None:
+        # This test is to ensure errors can propagate across futures.
+        error_msg = "Intentional Value Error"
+        value_error = ValueError(error_msg)
+
+        f = Future[T]()
+        # Set exception
+        f.set_exception(value_error)  # type: ignore
+        # Exception should throw on wait
+        with self.assertRaisesRegex(ValueError, "Intentional"):
+            f.wait()
+
+        # Exception should also throw on value
+        f = Future()
+        f.set_exception(value_error)  # type: ignore
+        with self.assertRaisesRegex(ValueError, "Intentional"):
+            f.value()  # type: ignore
```

Comment:
what happens when calling `then()`? I assume it should still work, and the cb can retrieve the exception from the `value`/`wait` API as well?

Reply:
It looks like the exception is only raised if we use value/wait in the then() code, otherwise the cb runs fine, which makes sense, since it doesn't use the value of the future. 

Due to https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/pybind_utils.h#L136, its RuntimeError instead of the specific error though.